meta/rpc.go

```diff
@@ -0,0 +1,430 @@
+		}
+		return ms, nil
+	case *internal.ErrorResponse:
+		return nil, fmt.Errorf("rpc failed: %s", t.GetHeader().GetError())
+	default:
+		return nil, fmt.Errorf("rpc failed: unknown response type: %v", t.String())
+	}
+}
+
+// join attempts to join a cluster at remoteAddr using localAddr as the current
+// node's cluster address
+func (r *RPC) join(localAddr, remoteAddr string) (*JoinResult, error) {
+	req := &internal.JoinRequest{
+		Addr: proto.String(localAddr),
+	}
+
+	resp, err := r.call(remoteAddr, req)
+	if err != nil {
+		return nil, err
+	}
+
+	switch t := resp.(type) {
+	case *internal.JoinResponse:
+		return &JoinResult{
+			RaftEnabled: t.GetEnableRaft(),
+			Peers:       t.GetPeers(),
+			NodeID:      t.GetNodeID(),
+		}, nil
+	case *internal.ErrorResponse:
+		return nil, fmt.Errorf("rpc failed: %s", t.GetHeader().GetError())
+	default:
+		return nil, fmt.Errorf("rpc failed: unknown response type: %v", t.String())
+	}
+}
+
+// call sends an encoded request to the remote leader and returns
+// an encoded response value.
+func (r *RPC) call(dest string, req proto.Message) (proto.Message, error) {
+
+	// Determine type of request
+	var rpcType internal.RPCType
+	switch t := req.(type) {
+	case *internal.JoinRequest:
+		rpcType = internal.RPCType_Join
+	case *internal.FetchDataRequest:
+		rpcType = internal.RPCType_FetchData
+	default:
+		return nil, fmt.Errorf("unknown rpc request type: %v", t)
+	}
+
+	// Create a connection to the leader.
+	conn, err := net.DialTimeout("tcp", dest, 10*time.Second)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	// Write a marker byte for rpc messages.
+	_, err = conn.Write([]byte{MuxRPCHeader})
+	if err != nil {
+		return nil, err
+	}
+
+	b, err := proto.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("rpc marshal: %v", err)
+	}
+
+	// Write request size & bytes.
+	if _, err := conn.Write(r.pack(rpcType, b)); err != nil {
+		return nil, fmt.Errorf("write %v rpc: %s", rpcType, err)
+	}
+
+	data, err := ioutil.ReadAll(conn)
+	if err != nil {
+		return nil, fmt.Errorf("read %v rpc: %v", rpcType, err)
+	}
+
+	// Should always have a size and type
+	if exp := 16; len(data) < exp {
+		return nil, fmt.Errorf("rpc %v failed: short read: got %v, exp %v", rpcType, len(data), exp)
+	}
+
+	sz := btou64(data[0:8])
+	if len(data[8:]) != int(sz) {
+		return nil, fmt.Errorf("rpc %v failed: short read: got %v, exp %v", rpcType, len(data[8:]), sz)
+	}
+
+	// See what response type we got back, could get a general error response
+	rpcType = internal.RPCType(btou64(data[8:16]))
+	data = data[16:]
+
+	var resp proto.Message
+	switch rpcType {
+	case internal.RPCType_Join:
+		resp = &internal.JoinResponse{}
+	case internal.RPCType_FetchData:
+		resp = &internal.FetchDataResponse{}
+	case internal.RPCType_Error:
+		resp = &internal.ErrorResponse{}
+	default:
+		return nil, fmt.Errorf("unknown rpc response type: %v", rpcType)
+	}
+
+	if err := proto.Unmarshal(data, resp); err != nil {
```

Comment:
What will happen if this `data` is now empty after being truncated by `data = data[16:]`?